### PR TITLE
Add `evaluate_demo` and `range_analysis_demo` to Expr examples

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/7013

## Rationale for this change

As part of updating DataFusion internally in IOx https://github.com/influxdata/influxdb_iox/pull/9428 , I found that we had code that used the range analysis code directly and thus had to be changed due to https://github.com/apache/arrow-datafusion/pull/8276

I found that DataFusion has a much nicer interface to do this [`analyze`](https://docs.rs/datafusion/latest/datafusion/physical_expr/fn.analyze.html) but I didn't think it was particularly obvious how to do so

It also came up on https://github.com/apache/arrow-datafusion/pull/8306#discussion_r1403489719 that it was non trivial to figure out how to evaluate expressions, so I added an example of how to do that as well

## What changes are included in this PR?

1. Add example of using analysis APIs 
2. Add example of evaluating an Expr, which is not as obvious either

### CODE REVIEW

File: datafusion-examples/examples/expr_api.rs [hunk 1]
```diff
@@ -120,6 +171,64 @@ fn simplify_demo() -> Result<()> {
...
+        lit(ScalarValue::Date32(Some(18506)))
+    );
+
+    Ok(())
+}
+
+/// DataFusion also has APIs for analyzing predicates (boolean expressions) to
+/// determine any ranges restrictions on the inputs required for the predicate
+/// evaluate to true.
+fn range_analysis_demo() -> Result<()> {
```
Comment (alamb) [inline datafusion-examples/examples/expr_api.rs:187]: This is a really powerful feature of DataFusion and I don't think it is widely understood yet

File: datafusion/physical-expr/src/analysis.rs [hunk 2]
```diff
@@ -111,6 +115,22 @@ impl ExprBoundaries {
             distinct_count: col_stats.distinct_count.clone(),
         })
     }
+
+    /// Create `ExprBoundaries` that represent no known bounds for all the columns `schema`
+    pub fn try_new_unknown(schema: &Schema) -> Result<Vec<Self>> {
```
Comment (alamb) [inline datafusion/physical-expr/src/analysis.rs:120]: This was added to make the demo easier to write (I ported it from upstream)

**Comment by alamb** [comment]
cc @ozankabak  / @berkaysynnada  / @metesynnada  as this PR adds examples to some of the great features you have added

**Comment by berkaysynnada** [comment]
These are really good demos. The utility functions you've added also make sense to me. Thanks, @alamb.

File: datafusion-examples/examples/expr_api.rs [hunk 3]
```diff
@@ -120,6 +171,64 @@ fn simplify_demo() -> Result<()> {
...
+    let expr = col("date")
+        .gt(lit(september_1.clone()))
+        .and(col("date").lt(lit(october_1.clone())));
+
+    // Using the analysis API, DataFusion can determine that the value of `date`
+    // must be in the range `['2020-09-01', '2020-10-01']`. If your data is
+    // organized in files according to day, this information permits skipping
+    // entire files without reading them.
+    //
+    // While this simply example could be handled with a special case, the
```
Comment (tustvold) [inline datafusion-examples/examples/expr_api.rs:204]:
```suggestion
    // While this simple example could be handled with a special case, the
```

File: datafusion/physical-expr/src/analysis.rs (hunk 2 above)
Comment (tustvold) [reply datafusion/physical-expr/src/analysis.rs:120]: Might unbounded be more obvious a name than unknown?

**Review by tustvold — APPROVED** [review]

File: datafusion/physical-expr/src/analysis.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/analysis.rs:120]: > Might unbounded be more obvious a name than unknown?

I agree -- will change.

**Pull request merged**